src/js_tests/wirecloud/ConnectionSpec.js

```diff
@@ -0,0 +1,251 @@
+
+            it("should create a new instance with no options", function () {
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+
+                expect(connection instanceof ns.Connection).toBe(true);
+                expect(connection.source).toBe(sourceEndpoint);
+                expect(connection.target).toBe(targetEndpoint);
+                expect(connection.readonly).toBe(false);
+                expect(connection.volatile).toBe(false);
+                expect(connection.id).toEqual([sourceEndpoint.id, targetEndpoint.id].join("//"));
+            });
+
+            it("should create a new instance with options.readonly = true", function () {
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint, {
+                    readonly: true
+                });
+
+                expect(connection.readonly).toBe(true);
+            });
+        });
+
+        describe("toJSON()", function () {
+
+            it("should parse to JSON", function () {
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+
+                expect(JSON.parse(JSON.stringify(connection))).toEqual({
+                    readonly: false,
+                    source: sourceEndpoint.toJSON(),
+                    target: targetEndpoint.toJSON()
+                });
+            });
+        });
+
+        describe("equals(value)", function () {
+
+            it("should be equal", function () {
+                var connection1 = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+                var connection2 = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+
+                expect(connection1.equals(connection2)).toBe(true);
+            });
+        });
+
+        describe("establish()", function () {
+
+            it("should connect two endpoints properly", function () {
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("establish", callback);
+                connection.establish();
+
+                expect(connection.established).toBe(true);
+                expect(callback.calls.count()).toEqual(1);
+            });
+
+            it("should ignore if connection is established", function () {
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("establish", callback);
+                connection.establish();
+                connection.establish();
+
+                expect(connection.established).toBe(true);
+                expect(callback.calls.count()).toEqual(1);
+            });
+
+            it("should ignore if one of their endpoints is missing", function () {
```

Comment:
Split into two tests:

"should do nothing if the source is a ghost endpoint"
"should do nothing if the target is a ghost endpoint"